plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java

```diff
@@ -1696,32 +1701,34 @@ public Optional<ConstraintApplicationResult<ConnectorTableHandle>> applyFilter(C
-
-        TupleDomain<IcebergColumnHandle> newEnforcedConstraint = constraint.getSummary()
-                .transformKeys(IcebergColumnHandle.class::cast)
-                .filter(isIdentityPartition)
-                .intersect(table.getEnforcedPredicate());
-
-        TupleDomain<IcebergColumnHandle> remainingConstraint = constraint.getSummary()
-                .transformKeys(IcebergColumnHandle.class::cast)
-                .filter(isIdentityPartition.negate());
-
-        TupleDomain<IcebergColumnHandle> newUnenforcedConstraint = remainingConstraint
-                // TODO: Remove after completing https://github.com/trinodb/trino/issues/8759
-                // Only applies to the unenforced constraint because structural types cannot be partition keys
-                .filter((columnHandle, predicate) -> !isStructuralType(columnHandle.getType()))
-                .filter(isMetadataColumn.negate())
-                .intersect(table.getUnenforcedPredicate());
+        Map<IcebergColumnHandle, Domain> unsupported = new LinkedHashMap<>();
+        Map<IcebergColumnHandle, Domain> newEnforced = new LinkedHashMap<>();
+        Map<IcebergColumnHandle, Domain> newUnenforced = new LinkedHashMap<>();
+        Map<ColumnHandle, Domain> domains = constraint.getSummary().getDomains().orElseThrow(() -> new IllegalArgumentException("constraint summary is NONE"));
+        domains.forEach((column, domain) -> {
+            IcebergColumnHandle columnHandle = (IcebergColumnHandle) column;
+            // Iceberg metadata columns can not be used to filter a table scan in Iceberg library
+            // TODO (https://github.com/trinodb/trino/issues/8759) structural types cannot be used to filter a table scan in Iceberg library.
+            if (isMetadataColumnId(columnHandle.getId()) || isStructuralType(columnHandle.getType())) {
+                unsupported.put(columnHandle, domain);
+            }
+            else if (canEnforceColumnConstraintInAllSpecs(typeOperators, icebergTable, columnHandle, domain)) {
+                newEnforced.put(columnHandle, domain);
+            }
+            else {
+                newUnenforced.put(columnHandle, domain);
+            }
+        });
+
+        TupleDomain<IcebergColumnHandle> newEnforcedConstraint = TupleDomain.withColumnDomains(newEnforced).intersect(table.getEnforcedPredicate());
+        TupleDomain<IcebergColumnHandle> newUnenforcedConstraint = TupleDomain.withColumnDomains(newUnenforced).intersect(table.getUnenforcedPredicate());
 
         if (newEnforcedConstraint.equals(table.getEnforcedPredicate())
                 && newUnenforcedConstraint.equals(table.getUnenforcedPredicate())) {
             return Optional.empty();
         }
 
+        TupleDomain<IcebergColumnHandle> remainingConstraint = TupleDomain.withColumnDomains(newUnenforced).intersect(TupleDomain.withColumnDomains(unsupported));
```

Comment:
I am probably missing something but in what scenario this intersection won't be empty? 
`newUnenforced` and `unsupported` are created and updated in the same if in different branches(few lines above) so how come they can ever intersect?

Reply:
@homar  I guess this a question what `TupleDomain.intersect` does.
`TupleDomain` technically, it's a Map, but should not be viewed as a collection. Rather, it's a constraint, a predicate. 
`TupleDomain.intersect` ANDs the predicates. If the predicates are defined on distinct sets of columns (supported vs not supported), that's not a problem.